### PR TITLE
PP-524: add travis CI support for debian

### PR DESCRIPTION
**Pull request by honza801**
#### Issue-ID

* **[PP-524](https://pbspro.atlassian.net/browse/PP-524)**

#### Problem description
* add travis CI support for debian

#### Cause / Analysis
* travis automatic builds and tests include only centos and suse environments. please add debian support.

#### Solution description
* split travis config into distribution specific scripts
* add travis CI support for debian
* remove docker 1.9 hack, v 1.12 works fine this way

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: .travis/deb.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash -x
+BUILDPKGS='build-essential autoconf libtool libssl-dev libxt-dev libpq-dev libexpat1-dev libedit-dev libncurses5-dev libical-dev libhwloc-dev pkg-config tcl-dev tk-dev python-dev swig'
+DEPPKGS='postgresql'
+TESTPKGS='python-pip sudo'
+[ "${OS_TYPE}" == "debian:8" ] && ${DOCKER_EXEC} apt-get -qq update || true
+[ "${OS_TYPE}" == "debian:8" ] && ${DOCKER_EXEC} apt-get install -y $BUILDPKGS $DEPPKGS $TESTPKGS || true
+${DOCKER_EXEC} ./autogen.sh
+${DOCKER_EXEC} ./configure
+${DOCKER_EXEC} make
+${DOCKER_EXEC} make install
```
Comment (mike0042) [inline .travis/deb.sh]: We should be creating a .deb package as opposed to running "make install". We may want to use the alien tool to accomplish this.

File: .travis.yml [hunk 2]
```diff
@@ -18,26 +16,9 @@ before_install:
...
-    - ${DOCKER_EXEC} /bin/sh -c 'cp -fv pbspro-*.tar.gz /root/rpmbuild/SOURCES/'
-    - ${DOCKER_EXEC} rpmbuild -bb pbspro.spec
-    - '[ "${OS_TYPE}" == "centos:7" ] && travis_retry ${DOCKER_EXEC} /bin/sh -c ''yum -y install /root/rpmbuild/RPMS/x86_64/pbspro-server-*.x86_64.rpm'' || true'
-    - '[ "${OS_TYPE}" == "opensuse:13.2" ] && travis_retry ${DOCKER_EXEC} /bin/sh -c ''zypper -n install /root/rpmbuild/RPMS/x86_64/pbspro-server-*.x86_64.rpm'' || true'
-    - ${DOCKER_EXEC} /etc/init.d/pbs start
-    - '[ "${OS_TYPE}" == "centos:7" ] && travis_retry ${DOCKER_EXEC} yum -y install python-pip sudo which net-tools || true'
-    - '[ "${OS_TYPE}" == "opensuse:13.2" ] && travis_retry ${DOCKER_EXEC} zypper -n install python-pip sudo which net-tools || true'
+    - '[ "${OS_TYPE}" == "centos:7" ] && travis_retry ./.travis/rpm.sh || true'
+    - '[ "${OS_TYPE}" == "opensuse:13.2" ] && travis_retry ./.travis/rpm.sh || true'
+    - '[ "${OS_TYPE}" == "debian:8" ] && travis_retry ./.travis/deb.sh || true'
```
Comment (mike0042) [inline .travis.yml]: I like that you broke these out into their own scripts, but it seems like there should be a script for each distro. There are still a lot of conditionals in the rpm script that could be eliminated if there were independent scripts for CentOS and openSUSE.

**Comment by honza801** [comment]
hi,

is the alien tool the preferred way over standard DEBIAN packaging tools?

may i split rpm/deb into debian/centos/opensuse or debian:8/centos:7/opensuse:13.2 ?

fous

**Comment by hirenvadalia** [comment]
@honza801 I think we need to add error check at each step in all distro script because Travis should fail immediately if any step in script fails.

**Comment by mike0042** [comment]
I think alien would be the correct approach in this case. That way we don't have to maintain two sets of packaging tools in the future. I'm no expert in Debian, so please feel free to disagree if I'm missing something.

**Comment by subhasisb** [comment]
I agree. Alien seems to be a better approach than creating a whole bunch of almost-duplicate files to support CI on debian.

**Comment by honza801** [comment]
building on debian using rpmbuild, ends with

Checking for unpackaged file(s): /usr/lib/rpm/check-files /root/rpmbuild/BUILDROOT/pbspro-14.0.1-0.x86_64
error: Installed (but unpackaged) file(s) found:
   /usr/lib/systemd/system/pbs.service

which is probably caused by conditional
%if 0%{?suse_version} >= 1210 || 0%{?rhel} >= 7
%define have_systemd 1
%endif

as i'm not an expert on rpm spec files. is there any way to determine that i'm building package on debian, or get the debian distribution version?

**Comment by hirenvadalia** [comment]
@honza801 by "error check at each step" means travis scripts for all dist should fail on first error which you already handled by added "bash -xe".
And also you are right about "|| true" removal as it will disable error check. Can you please remove that from scripts?

**Comment by subhasisb** [comment]
@honza801 please update the code based on request from @hirenvadalia - we will be able to merge it in soon after that.

**Comment by honza801** [comment]
ok, done.

**Comment by hirenvadalia** [comment]
@honza801 Thank you for making changes.
One more request for making changes (sorry for not catching this earlier):
Since ':' is invalid character in Windows we have to change name of all dist files in .travis directory otherwise git will delete those file and create new files with name like centos, debian and opensuse with no content in it (see below) in Windows:
```
$ git fetch origin pull/202/head:pr-202
remote: Counting objects: 13, done.
remote: Compressing objects: 100% (12/12), done.
remote: Total 13 (delta 2), reused 1 (delta 1), pack-reused 0
Unpacking objects: 100% (13/13), done.
From https://github.com/PBSPro/pbspro
 * [new ref]         refs/pull/202/head -> pr-202

$ git checkout pr-202
Switched to branch 'pr-202'

$ git status
On branch pr-202
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        deleted:    .travis/centos:7.sh
        deleted:    .travis/debian:8.sh
        deleted:    .travis/opensuse:13.2.sh

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        .travis/centos
        .travis/debian
        .travis/opensuse

no changes added to commit (use "git add" and/or "git commit -a")
$ cat .travis/centos

$ cat .travis/debian

$ cat .travis/opensuse

$ ls -la .travis
total 5
drwxr-xr-x 1 hiren 1049089  0 Nov 30 19:35 ./
drwxr-xr-x 1 hiren 1049089  0 Nov 30 19:35 ../
-rw-r--r-- 1 hiren 1049089  0 Nov 30 19:35 centos
-rw-r--r-- 1 hiren 1049089  0 Nov 30 19:35 debian
-rw-r--r-- 1 hiren 1049089 73 Nov 30 19:35 debian.macros.dist
-rw-r--r-- 1 hiren 1049089  0 Nov 30 19:35 opensuse

```

Suggested solution:
* Remove ':' from file names e.g. centos:7.sh will become centos7.sh
* In .travis.yml replace 'travis_retry ./.travis/${OS_TYPE}.sh' with 'travis_retry ./.travis/${OS_TYPE//[:]}.sh'

**Comment by honza801** [comment]
i see.

i think there should be some version separator. i have used 'travis_retry ./.travis/${OS_TYPE//[:]/_}.sh'

**Comment by subhasisb** [comment]
Awesome! Thanks@Mike

**Pull request merged**